src/scroll-timeline-css-parser.js

```diff
@@ -0,0 +1,381 @@
+  '%',
+];
+
+export const RegexMatcher = {
+  IDENTIFIER: /[\w\\\@_-]+/g,
+  WHITE_SPACE: /\s*/g,
+  NUMBER: /^[0-9]+/,
+  TIME: /^[0-9]+(s|ms)/,
+  ANIMATION_TIMELINE: /animation-timeline\s*:([^;}]+)/,
+  ANIMATION_NAME: /animation-name\s*:([^;}]+)/,
+  ANIMATION: /animation\s*:([^;}]+)/,
+  OFFSET_WITH_SUFFIX: new RegExp('(^[0-9]+)(' + VALID_SCROLL_OFFSET_SUFFIXES.join('|') + ')'),
+  ELEMENT_OFFSET: /selector\(#([^)]+)\)[ ]{0,1}(start|end)*[ ]{0,1}([0-9]+[.]{0,1}[0-9]*)*/,
+  SOURCE_ELEMENT: /selector\(#([^)]+)\)/,
+};
+
+// Used for ANIMATION_TIMELINE, ANIMATION_NAME and ANIMATION regex
+const VALUES_CAPTURE_INDEX = 1;
+
+const WHOLE_MATCH_INDEX = 0;
+
+const ANIMATION_KEYWORDS = [
+  'normal', 'reverse', 'alternate', 'alternate-reverse',
+  'none', 'forwards', 'backwards', 'both',
+  'running', 'paused',
+  'ease', 'linear', 'ease-in', 'ease-out', 'ease-in-out'
+];
+
+// 1 - Extracts @scroll-timeline and saves it in scrollTimelineOptions.
+// 2 - If we find any animation-timeline in any of the CSS Rules, 
+// we will save objects in a list named cssRulesWithTimelineName
+export class StyleParser {
+  constructor() {
+    this.cssRulesWithTimelineName = [];
+    this.scrollTimelineOptions = new Map(); // save options by name
+    this.keyframeNames = new Set();
+  }
+
+  // Inspired by
+  // https://drafts.csswg.org/css-syntax/#parser-diagrams
+  // https://github.com/GoogleChromeLabs/container-query-polyfill/blob/main/src/engine.ts
+  // This function is called twice, in the first pass we are interested in saving
+  // @scroll-timeline and @keyframe names, in the second pass
+  // we will parse other rules
+  transpileStyleSheet(sheetSrc, firstPass, srcUrl) {
+    // AdhocParser
+    const p = {
+      sheetSrc: sheetSrc,
+      index: 0,
+      name: srcUrl,
+    };
+
+    while (p.index < p.sheetSrc.length) {
+      this.eatWhitespace(p);
+      if (p.index >= p.sheetSrc.length) break;
+      if (this.lookAhead("/*", p)) {
+        while (this.lookAhead("/*", p)) {
+          this.eatComment(p);
+          this.eatWhitespace(p);
+        }
+        continue;
+      }
+
+      if (this.lookAhead("@scroll-timeline", p)) {
+        const { scrollTimeline, startIndex, endIndex } = this.parseScrollTimeline(p);
+        if (firstPass) this.scrollTimelineOptions.set(scrollTimeline.name, scrollTimeline);
+      } else {
+        const rule = this.parseQualifiedRule(p);
+        if (!rule) continue;
+        if (firstPass) {
+          this.extractAndSaveKeyframeName(rule.selector);
+        } else {
+          this.handleScrollTimelineProps(rule, p);
+        }
+      }
+    }
+
+    // If this sheet has no srcURL (like from a <style> tag), we are done.
+    // TODO: Otherwise, we have to find `url()` functions and resolve
+    // relative and path-absolute URLs to absolute URLs.
+    return p.sheetSrc;
+  }
+
+  getScrollTimelineName(animationName, target) {
+    for (let i = this.cssRulesWithTimelineName.length - 1; i >= 0; i--) {
```

Comment:
Add comment why we are traversing backwards along with TODO for rule specificity.